proposals/csharp-12.0/collection-expressions.md

```diff
@@ -255,6 +255,52 @@ ImmutableArray<int> ia =
     List<int> z = [];
     ```
 
+## Spreads
+[spreads]: #spreads
+
+```csharp
+bool b = true;
+int[] x = [.. []];            // x = []
+byte[] y = [.. b ? [1] : []]; // y = [(byte)1]
+```
+
+A *spread element* `..E` is applicable as an iterator of type `T` if any of the following hold:
+
+* There is an implicit conversion from the [*iteration type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) of `E` to `T`.
+* `E` is a *collection expression* and for each element `Eᵢ` there is an implicit conversion to `T`.
+* `E` is a [*conditional expression*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1115-conditional-operator) `b ? x : y` where one or both of `x` and `y` is a *collection expression* and for each of the operands `x` and `y` one of the following hold:
+  * The operand is a collection expression and for each element `Eᵢ` there is an implicit conversion to `T`.
+  * The operand has an [*iteration type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) `Tₑ` and there is an implicit conversion from `Tₑ` to `T`.
+
+For a collection expression used a spread element expression, or for a collection expression as the second or third operand in a conditional expression used as a spread element expression, the compiler may use any conforming representation for the collection instance, including eliding the collection.
+
+For example, the code generated for `List<int> list = [x, y, .. b ? [z] : []]` could avoid an intermediate collection:
+```c#
+List<int> list = new();
+list.Add(x);
+list.Add(y);
+if (b) list.Add(z);
+```
+
+*Should `switch` expressions be supported in spread elements?*
+
+## Foreach
+[foreach]: #foreach
+
+```csharp
+foreach (bool? b in [false, true, null]) { ... }
+```
+
+A *collection expression* may be used as the collection in a `foreach` statement with an *explicitly typed iteration variable* of type `T` if:
+
+* For each element `Eᵢ` in the collection expression there is an implicit conversion to `T`.
+
+An error is reported if the `foreach` iteration variable is *implicitly typed*.
```

Comment:
Rephrasing @CyrusNajmabadi's question here: Can we make this work with `var b in`, without invoking natural typing, the same way `Foreach<T>(ROS<T>)` would be called without natural typing?

- And maybe, permitting ref structs to be used? `foreach (var x in [refStruct1, refStruct2])`

Reply:
That would be great too :)

Reply:
> * And maybe, permitting ref structs to be used? `foreach (var x in [refStruct1, refStruct2])`

Is it possible to also allow `ref struct` in tuple being deconstructed ?
`(Prop1, Prop2) = (new RefStruct(), new RefStruct())`
[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCecAmIA1AD4ACATAIwCwAUGQMwAE6qWzAzgugK4DGCZgGVmAb3rMpzJiJ4ChAQTjMAQgG5J02QCV2w+YOYBhFQBFNdaTJZkALMwCyACgCU4rdanPla9wF5mZxhUAHc5PkE3FRDwg0iEN1dLL2lnU2YzAKDY5j0seIVo5lz8wqjXZM9mAF96Orp6WTYObgS8/UMhMVr6JpY2hQihnpqgA==)

_I know it's out of scope for this subject, but your question reminded me that_

Reply:
> Can we make this work with var b in, without invoking natural typing, the same way Foreach<T>(ROS<T>) would be called without natural typing?

We could use the [*best common type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#116315-finding-the-best-common-type-of-a-set-of-expressions) of the elements to determine the collection expression element type, which matches how the element type is determined for [*implicitly typed array creation expressions*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#117155-array-creation-expressions).